In [13]:
import numpy as np
import scipy 
import os
import matplotlib.pyplot as plot
import math
from scipy import ndimage

In [14]:
import nbimporter
from ShallowNN import * #import from notebook

## Test Cases for NN

### Initialize_params_shallow

In [15]:
np.random.seed(1)
params_shallow = initialize_params_shallow(3, 4, 2)
print("w1:", params_shallow["w1"])
print("w2:", params_shallow["w2"])
print("b1:", params_shallow["b1"])
print("b2:", params_shallow["b2"])

w1: [[ 0.01624345 -0.00611756 -0.00528172]
 [-0.01072969  0.00865408 -0.02301539]
 [ 0.01744812 -0.00761207  0.00319039]
 [-0.0024937   0.01462108 -0.02060141]]
w2: [[-0.00322417 -0.00384054  0.01133769 -0.01099891]
 [-0.00172428 -0.00877858  0.00042214  0.00582815]]
b1: [[ 0.]
 [ 0.]
 [ 0.]
 [ 0.]]
b2: [[ 0.]
 [ 0.]]


##### (3,4,2)
<table style:"width-80%">
    <tr>
        <td>
         <b> w1:</b> [[ 0.01624345 -0.00611756 -0.00528172]
         [-0.01072969  0.00865408 -0.02301539]
         [ 0.01744812 -0.00761207  0.00319039]
         [-0.0024937   0.01462108 -0.02060141]]
        </td>
        <td>
        w2: [[-0.00322417 -0.00384054  0.01133769 -0.01099891]
 [-0.00172428 -0.00877858  0.00042214  0.00582815]]
        </td>
    </tr>
    <tr>
    <td>
        b1: [[ 0.]
 [ 0.]
 [ 0.]
 [ 0.]]
        </td>
        <td>
b2: [[ 0.]
 [ 0.]]
 </td>
     </tr>
</table>

### Initialize_params_deep

In [26]:
np.random.seed(3)
params_deep = initialize_params_deep(5,[4],3)
for param in params_deep:
    print(param, params_deep[param].shape)
    print(params_deep[param])


w1 (4, 5)
[[ 0.01788628  0.0043651   0.00096497 -0.01863493 -0.00277388]
 [-0.00354759 -0.00082741 -0.00627001 -0.00043818 -0.00477218]
 [-0.01313865  0.00884622  0.00881318  0.01709573  0.00050034]
 [-0.00404677 -0.0054536  -0.01546477  0.00982367 -0.01101068]]
b1 (4, 1)
[[ 0.]
 [ 0.]
 [ 0.]
 [ 0.]]
w2 (3, 4)
[[-0.01185047 -0.0020565   0.01486148  0.00236716]
 [-0.01023785 -0.00712993  0.00625245 -0.00160513]
 [-0.00768836 -0.00230031  0.00745056  0.01976111]]
b2 (3, 1)
[[ 0.]
 [ 0.]
 [ 0.]]


##### seed(3) (5,[3],3)
<table style:"width-80%">
    <tr>
        <td>
w1 (4, 5)
[[ 0.01788628  0.0043651   0.00096497 -0.01863493 -0.00277388]
 [-0.00354759 -0.00082741 -0.00627001 -0.00043818 -0.00477218]
 [-0.01313865  0.00884622  0.00881318  0.01709573  0.00050034]
 [-0.00404677 -0.0054536  -0.01546477  0.00982367 -0.01101068]]
 </td>
 <td>
b1 (4, 1)
[[ 0.]
 [ 0.]
 [ 0.]
 [ 0.]]
</td>
<tr>
<td>
w2 (3, 4)
[[-0.01185047 -0.0020565   0.01486148  0.00236716]
 [-0.01023785 -0.00712993  0.00625245 -0.00160513]
 [-0.00768836 -0.00230031  0.00745056  0.01976111]]
</td>
<td>
b2 (3, 1)
[[ 0.]
 [ 0.]
 [ 0.]]
 </td>
 </tr>
 </table>

### Linear_foward_prop

In [17]:
np.random.seed(1)
X = np.random.randn(3,2)
w = np.random.randn(1,3)
b = np.random.randn(1) #(1,1) broadcasting
z,_ = linear_forward_prop(X,w,b)
print("X",X.shape,"w",w.shape,"b",b.shape,"z",z.shape)
print(z)

X (3, 2) w (1, 3) b (1,) z (1, 2)
[[ 3.26295337 -1.23429987]]


##### seed(1) x=(3,2) w=(1,3) b=(1)
<table style:"width-80%">
    <tr>
        <td>
X (3, 2) w (1, 3) b (1,) z (2, 1)
</td>
</tr>
    <tr>
        <td>
[[ 3.26295337]
 [-1.23429987]]
 </td>
</tr>
</table>

### SingleFowardProp

In [18]:
np.random.seed(2)
X = np.random.randn(3,2)
w = np.random.randn(1,3)
b = np.random.randn(1,1) #(1,1) broadcasting
a,cached_values = singleFowardPropagation(X,w,b,"relu")
print("Relu A",a)

a,cached_values = singleFowardPropagation(X,w,b,"sigmoid")
print("Sigmoud A",a)
# print(X,w,b)

Relu A [[ 3.43896131  0.        ]]
Sigmoud A [[ 0.96890023  0.11013289]]


Relu A [[ 3.43896131] [ 0. ]] Sigmoud A [[ 0.96890023] [ 0.11013289]]

In [19]:
np.random.seed(6)
X = np.random.randn(5,4)
w1 = np.random.randn(4,5)
b1 = np.random.randn(4,1)
w2 = np.random.randn(3,4)
b2 = np.random.randn(3,1)
w3 = np.random.randn(1,3)
b3 = np.random.randn(1,1)
params = {
    "w1":w1,
    "b1":b1,
    "w2":w2,
    "b2":b2,
    "w3":w3,
    "b3":b3
}
a, cached_values = fowardPropagate(X, params)
print("a:",a)
# print("cached_values",cached_values)

a: [[ 0.03921668  0.70498921  0.19734387  0.04728177]]


Before Transpose -- A: (5, 4) W: (4, 5) B: (4, 1) <br/>
After Transpose -- A: (5, 4) W: (4, 5) B: (4, 1) <br/>
z: (4, 4) <br/>
Before Transpose -- A: (4, 4) W: (3, 4) B: (3, 1) <br/>
After Transpose -- A: (4, 4) W: (3, 4) B: (3, 1) <br/>
z: (3, 4)<br/>
Before Transpose -- A: (3, 4) W: (1, 3) B: (1, 1) <br/>
After Transpose -- A: (3, 4) W: (1, 3) B: (1, 1) <br/>
z: (1, 4) <br/>
a: [[ 0.03921668  0.70498921  0.19734387  0.04728177]] <br/>

### LinearBackprop 

In [20]:
np.random.seed(1)
dZ = np.random.randn(1,2)
a = np.random.randn(3,2)
w = np.random.randn(1,3)
b = np.random.randn(1,1)
cached_values = {"w": w, "in_a":a,"b":b}
linearBackprop(cached_values, dZ)

{'dA': array([[ 0.51822968, -0.19517421],
        [-0.40506361,  0.15255393],
        [ 2.37496825, -0.89445391]]),
 'dB': array([[ 0.50629448]]),
 'dW': array([[-0.10076895,  1.40685096,  1.64992505]])}

<tr>
<td>
{'dA': array([[ 0.51822968, -0.19517421], [-0.40506361, 0.15255393], [ 2.37496825, -0.89445391]]),
</td>
<td>
'dB': array([[ 0.50629448]]), 'dW': array([[-0.10076895, 1.40685096, 1.64992505]])}
</td>
</tr>

### Backwards_activation

In [21]:
np.random.seed(2)
dA = np.random.randn(1,2)
a = np.random.randn(3,2)
w = np.random.randn(1,3)
b = np.random.randn(1,1)
z = np.random.randn(1,2)
cached_values = {"w": w, "in_a":a,"b":b, "z":z}
backwards_activation(dA, cached_values, activation_type="relu")
backwards_activation(dA, cached_values, activation_type="sigmoid")

{'dA': array([[ 0.11017994,  0.01105339],
        [ 0.09466817,  0.00949723],
        [-0.05743092, -0.00576154]]),
 'dB': array([[-0.05729622]]),
 'dW': array([[ 0.10266786,  0.09778551, -0.01968084]])}

<table style:"width-80%">
<tr>
<td>
For Relu: {'dA': array([[ 0.44090989, 0. ], [ 0.37883606, 0. ], [-0.2298228 , 0. ]]), 'dB': array([[-0.20837892]]), 'dW': array([[ 0.44513824, 0.37371418, -0.10478989]])}
</td>
</tr>
<tr>
<td>
For Sigmoid: {'dA': array([[ 0.11017994, 0.01105339], [ 0.09466817, 0.00949723], [-0.05743092, -0.00576154]]), 'dB': array([[-0.05729622]]), 'dW': array([[ 0.10266786, 0.09778551, -0.01968084]])}
</td>
</tr>
</table>

### Backprop

In [22]:
np.random.seed(3)
a = np.random.randn(1, 2)
a1 = np.random.randn(4, 2)
b1 = np.random.randn(3, 4)
w1 = np.random.randn(3, 1)
z1 = np.random.randn(3, 2)
cache1 = [{"in_a" : a1,"b" : b1,"w" : w1, "z" : z1}]
a2 = np.random.randn(3, 2)
b2 = np.random.randn(1, 1)
w2 = np.random.randn(1, 3)
z2 = np.random.randn(1, 2)
cache2 = [{"in_a" : a2,"b" : b2,"w" : w2, "z" : z2}]
megaCache = cache1 +  cache2
y = np.array([[1,0]])
# print("y",y)
# print(megaCache[0]["in_a"],"in_a1")
# print(megaCache[0]["b"],"b1")
# print(megaCache[0]["w"],"w1")
# print(megaCache[0]["z"],"z1")
# print(megaCache[1]["in_a"],"in_a2")
# print(megaCache[1]["b"],"b2")
# print(megaCache[1]["w"],"w2")
# print(megaCache[1]["z"],"z2")


grads = backprop(a, megaCache,y)
print("dA0", grads["dA0"].shape)
print("dW1", grads["dW1"].shape)
print("dB1", grads["dB1"].shape)
print("dW2", grads["dW2"].shape)
print("dB2", grads["dB2"].shape)
print("dA1", grads["dA1"].shape)
print(grads)

dA0 (1, 2)
dW1 (3, 4)
dB1 (3, 1)
dW2 (1, 3)
dB2 (1, 1)
dA1 (3, 2)
{'dW2': array([[-0.39202432, -0.13325855, -0.04601089]]), 'dB2': array([[ 0.15187861]]), 'dA1': array([[-0.14175655,  0.48317296],
       [ 0.01663708, -0.05670698],
       [ 0.20472941, -0.69781407]]), 'dW1': array([[-0.45019464, -0.08570497, -0.15147489, -0.11528942],
       [ 0.        ,  0.        ,  0.        ,  0.        ],
       [ 0.65018572,  0.1237779 ,  0.21876495,  0.16650473]]), 'dB1': array([[ 0.24158648],
       [ 0.        ],
       [-0.34890704]]), 'dA0': array([[ 0.        ,  1.43247838]])}


dA0 (1, 2)
dW1 (3, 4)
dB1 (3, 1)
dW2 (1, 3)
dB2 (1, 1)
dA1 (3, 2)
{'dW2': array([[-0.39202432, -0.13325855, -0.04601089]]), 'dB2': array([[ 0.15187861]]), 'dA1': array([[-0.14175655,  0.48317296],
       [ 0.01663708, -0.05670698],
       [ 0.20472941, -0.69781407]]), 'dW1': array([[-0.45019464, -0.08570497, -0.15147489, -0.11528942],
       [ 0.        ,  0.        ,  0.        ,  0.        ],
       [ 0.65018572,  0.1237779 ,  0.21876495,  0.16650473]]), 'dB1': array([[ 0.24158648],
       [ 0.        ],
       [-0.34890704]]), 'dA0': array([[ 0.        ,  1.43247838]])}